src/hil/spi_master.rs

```diff
@@ -1,73 +1,86 @@
-//! Traits and parameters for SPI master communication
+//! Traits and parameters for SPI master communication.
+//! Authors: Sam Crow <samcrow@uw.edu>
+//!          Philip Levis <pal@cs.stanfored.edu>
 
 use core::option::Option;
 
-#[derive(Copy, Clone)]
-pub enum Rate {
-    MSBFirst,
-    LSBFirst,
-}
 /// Values for the ordering of bits
 #[derive(Copy, Clone)]
-pub enum DataOrder {
-    MSBFirst,
-    LSBFirst,
-}
+pub enum DataOrder {MSBFirst, LSBFirst}
 
 /// Values for the clock polarity (idle state or CPOL)
 #[derive(Copy, Clone)]
-pub enum ClockPolarity {
-    IdleHigh,
-    IdleLow,
-}
-/// Values for the clock phase (CPHA), which defines when
-/// values are sampled
+pub enum ClockPolarity {IdleHigh, IdleLow}
+
+/// Which clock edge values are sampled on
 #[derive(Copy, Clone)]
-pub enum ClockPhase {
-    SampleLeading,
-    SampleTrailing,
-}
+pub enum ClockPhase {SampleLeading, SampleTrailing}
 
 pub trait SpiCallback {
     /// Called when a read/write operation finishes
-    fn read_write_done(&'static self, 
-                       read: Option<&'static mut[u8]>,
-                       writer: Option<&'static mut[u8]>);
```

Comment:
I thought we discussed this earlier. The problem with passing AppSlice is that you can't terminate the application while the call is outstanding. This gives us 2 options:

  1) Keep track of outstanding operations and terminate application only when all of them have completed
  2) Have drivers copy into/out of an in-kernel buffer ('static), which then is used to invoke HAL

I think #1 is better long-term but it's going to require some careful design to get right. So I'm assuming I'm going with #2 for now.